pennylane/math/__init__.py

```diff
@@ -109,6 +110,29 @@ def __getattribute__(self, fn):
         return super().__getattribute__(fn)
 
 
+def ndim(a):
+    '''Custom implementation of autoray.ndim to avoid overheads when calling this
+    method many times in the parameter shift rule.'''
+    if isinstance(a, autograd.numpy.numpy_boxes.ArrayBox):
+        return autograd.numpy.ndim(a)
```

Comment:
If we stick to the `try`-`except` solution we can remove this, since `ArrayBox`es have `ndim` and `shape` attributes.